doc/design/milestone-0.2.0/csv-generation.md

```diff
@@ -0,0 +1,176 @@
+
+The `operator-sdk generate olm-catalog` command currently produces a generic CSV with minimal customization. Defaults and simple metadata components are used to fill fields, with no options to customize. Users must modify the CSV manually or use custom scripts to pull data from various operator component files. These solutions are not scalable because we cannot assume users are, or have time to become, familiar with CSV format or know where to find information required by CSV yaml fields.
+
+## Proposed Solution
+
+Functionality of `operator-sdk generate olm-catalog` is now a branch in `operator-sdk build`, specifically when called with the `--gen-csv` flag; the former command no longer exists. `operator-sdk build quay.io/example/operator:v0.0.1 --gen-csv"` writes a CSV yaml file using the operators' current version to the `deploy/olm-catalog` directory by default. `deploy` is the standard location for all manifests and scripts required to deploy an operator. The SDK can use data from manifests in `deploy` to write a CSV. Users may have different requirements for what should (not) be included in a CSV, and can configure CSV composition via `deploy/olm-catalog/csv-config.yaml`, described below.
+
+Assuming all configuration defaults are used, `operator-sdk build` will call `scaffold.Execute()`, which will either:
+
+1. Create a new CSV, with the same location and naming convention as exists currently, using available data in yaml manifests and source files.
+
+    1. The update mechanism will check for an existing CSV in `deploy`. Upon not finding one, a [`ClusterServiceVersion` object][olm_csv_struct_code], referred to here as a *cache*, is created and fields easily derived from operator metadata, such as Kubernetes API `ObjectMeta`, are populated.
+    1. The update mechanism will search `deploy` for manifests that contain data a CSV uses, such as a `Deployment` Kubernetes API resource, and set the appropriate CSV fields in the cache with this data.
+    1. Once the search completes, every cache field populated will be written back to a CSV yaml file.
+        - **Note:** individual yaml fields are overwritten and not the entire file, as descriptions and other non-generated parts of a CSV should be preserved.
+
+1. Update an existing CSV at the currently pre-defined location, using available data in yaml manifests and source files.
+
+    1. The update mechanism will check for an existing CSV in `deploy`. Upon finding one, the CSV yaml file contents will be marshalled into a `ClusterServiceVersion` cache.
+    1. Same as above.
+    1. Same as above.
+
+### Configuration
+
+Users can configure CSV composition by populating several fields in the file `deploy/olm-catalog/csv-config.yaml`:
+
+- `search-directory`: (string) the directory to search for yaml manifest files. Defaults to `deploy`.
```

Comment:
More of an implementation detail but wanted to point out that since the CSV generator needs exactly 3 things to populate the CSV:
- The operator deployment manifest
- A CRD directory(with *_crd.yaml and *_cr.yaml) files
- The RBAC Role manifest with the rules defined

We can just specify those paths in the config file instead of a search directory. This way it's clear exactly what manifests are required, and it's flexible enough to let the user specify what manifests they want to input to the CSV generator. Also no need for `exclude-files` in that case I think.

If it's just a path to a search directory then I assume we would still expect some layout within the search directory similar to our default `deploy` directory layout. If not, then we're just searching the directory for what we need then that's more ambiguous.

Reply:
Should we assume multiple RBAC files will be present? I'm thinking of something like;
```Markdown
- `rbac-path-list`: ([string[, string]\*]) a list of paths to RBAC role manifest files. Defaults to `[deploy/role.yaml]`.
```

Reply:
No I think there should be only 1 RBAC Role manifest. 
Can't imagine a scenario where you'd bind two different Roles to a single service account. I'm not sure but that would just end up being a combination of the two.